plugins/blanker_control.py

```diff
@@ -0,0 +1,134 @@
+# -*- coding: utf-8 -*-
+"""
+Created on 29 Feb 2024
+
+@author: Éric Piel
+
+Gives ability to acquire a Spectrum scan with the e-beam position following an arbitrary order.
+The scan order has to be specified by a python function in this file.
+
+Copyright © 2024 Éric Piel, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms of the GNU
+General Public License version 2 as published by the Free Software Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
+the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+Public License for more details.
+
+You should have received a copy of the GNU General Public License along with Odemis. If not,
+see http://www.gnu.org/licenses/.
+"""
+import logging
+from typing import Any
+
+import wx
+
+from odemis import model
+from odemis.gui.plugin import Plugin
+from odemis.gui.util import call_in_wx_main
+
+# This plugin provides an extra menu entry Acquisition/E-beam blanker to force the blanker on/off
+# or put it back to automatic (ie, disabled only when the SEM stream is active).
+
+BLANKER_TO_MENU = {
+    None: "Automatic",
+    True: "Enabled",  # No e-beam
+    False: "Disabled",
+}
+MENU_PATH = "Acquisition/E-beam blanker"
+
+
+class BlankerControlPlugin(Plugin):
+    name = "E-beam Blanker Control"
+    __version__ = "1.0"
+    __author__ = "Éric Piel"
+    __license__ = "GPLv2"
+
+    def __init__(self, microscope, main_app):
+        super().__init__(microscope, main_app)
+        # Can only be used with a SPARC with spectrometer and CCD camera
+        main_data = self.main_app.main_data
+        if not main_data.ebeam or not model.hasVA("ebeam", "blanker"):
+            logging.info("%s plugin cannot load as the microscope has no e-beam blanker",
+                         self.name)
+
+        blanker = main_data.ebeam.blanker
+        # Add either a check menu if just on/off, or a radio menu if there are 3 states
+        try:
+            blanker_choices = blanker.choices
+        except AttributeError:
+            # No choices => it's probably just a BooleanVA
+            # There is no explicit way to check, so we just check whether the value is a boolean
+            if blanker.value not in {True, False}:
+                logging.warning("Unexpected value for e-beam blanker: %s", blanker.value)
+                return
+            blanker_choices = {True, False}
+
+        if len(blanker_choices) <= 1:
+            logging.info(f"E-beam blanker has only {len(blanker_choices)} state, no control menu added")
+            return
+        elif blanker_choices == {True, False}:  # Just on/off => make it a check menu
+            self._blanker_menu_item = self._addMenu(MENU_PATH, self._on_ebeam_blanker_bool,
```

Comment:
why are you calling `_addMenu` and not `addMenu`?

Reply:
because I need the return value. `addMenu` is the "safe" asynchronous version.